src/app/modules/ledger/ledger.service.js

```diff
@@ -0,0 +1,353 @@
+        }
+    }
+
+    async getRemoteAccount(hdKeypath) {
+        try {
+            const transport = await TransportNodeHid.open("");
+            const nemH = new NemH(transport);
+            try {
+                const result = await nemH.getRemoteAccount(hdKeypath)
+                transport.close();
+                return Promise.resolve(
+                    {
+                        "hdKeypath": hdKeypath,
+                        "remoteAccount": result.remoteAccount
+                    }
+                );
+            } catch (err) {
+                transport.close();
+                throw err
+            }
+        } catch (err) {
+            if (err.statusCode != null) return Promise.reject(err.statusCode);
+            else if (err.id != null) return Promise.reject(err.id);
+            else return Promise.reject(err);
+        }
+    }
+
+    deriveRemote(account, network) {
+        return new Promise((resolve, reject) => {
+            this.getAppVersion().then(checkVersion => {
+                if (checkVersion == 1) {
+                    alert("Please check your Ledger device!");
+                    this._$timeout(() => {
+                        this._Alert.ledgerFollowInstruction();
+                    });
+
+                    this.getRemoteAccount(account.hdKeypath).then((result) => {
+                        const privateKey = result.remoteAccount;
+                        const keyPair = nem.crypto.keyPair.create(privateKey);
+                        const publicKey = keyPair.publicKey.toString();
+                        const address = nem.model.address.toAddress(publicKey, network);
+                        resolve({
+                            address,
+                            privateKey,
+                            publicKey
+                        });
+                    }).catch((errorCode) => {
+                        this._$timeout(() => {
+                            this.alertHandler(errorCode);
+                        });
+                    });
+                } else {
+                    this._$timeout(() => {
+                        this.alertHandler(checkVersion);
+                    });
+                }
+            });
+        });
+    }
+
+    serialize(transaction, account, symbolOptin) {
+        return new Promise(async (resolve, reject) => {
+            this.getAppVersion().then(async checkVersion => {
+                if (checkVersion == 1) {
+                    alert("Please check your Ledger device!");
+                    this._$timeout(() => {
+                        this._Alert.ledgerFollowInstruction();
+                    });
+                    //Transaction with testnet and mainnet
+                    //Correct the signer
+                    transaction.signer = account.publicKey;
+
+                    //If it is a MosaicDefinition Creation Transaction, then correct the creator
+                    if (transaction.type == 0x4001) {
```

Comment:
Could you elaborate on why here to set the creator for mosaic?

Could could the serialization use `nem.model.transactions.prepare` in nem-sdk?

Reply:
Currently, the NEM-sdk hasn't support for hardware while preparing the transaction object yet when running the nem.model.transaction.prepare, for example when we create mosaic this function has no data of mosaic creator (ref [here](https://github.com/QuantumMechanics/NEM-sdk/blob/a9d93b1bced51c0c2c6838a01e233896da1dd847/src/model/transactions/mosaicDefinitionTransaction.js#L19-L21)) cause the wallet doesn't have private key of ledger device. So we have to correct the data of signer like [this](https://github.com/FDS-Ledger/NanoWallet/blob/d3734feceaa37fbc220317eb04d995fd29a62134/src/app/modules/ledger/ledger.service.js#L236-L242). Maybe some modification in NEM-sdk will fix it